src/CoreWCF.Http/src/CoreWCF/Channels/HttpChannelHelpers.cs

```diff
@@ -121,80 +123,62 @@ private void ThrowMaxReceivedMessageSizeExceeded()
-                }
+                return await _messageEncoder.ReadMessageAsync(buffer, _bufferManager, ContentType);
             }
-            finally
+            catch (XmlException xmlException)
             {
-                inputStream.Close();
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(
+                    new ProtocolException(SR.MessageXmlProtocolError, xmlException));
             }
         }
 
-        private async Task<Message> ReadBufferedMessageAsync(Stream inputStream)
+        private async Task<Message> ReadMessageAsync(PipeReader reader)
         {
-            ArraySegment<byte> messageBuffer = GetMessageBuffer();
-            byte[] buffer = messageBuffer.Array;
-            int offset = 0;
-            int count = messageBuffer.Count;
-
-            while (count > 0)
+            ReadOnlySequence<byte> buffer = ReadOnlySequence<byte>.Empty;
+            Message message = null;
+            try
             {
-                int bytesRead = await inputStream.ReadAsync(buffer, offset, count);
-                if (bytesRead == 0) // EOF
+                while (true)
                 {
-                    if (ContentLength != -1)
+                    ReadResult result = await reader.ReadAsync();
+                    buffer = result.Buffer;
+
+                    if (buffer.Length >= _settings.MaxReceivedMessageSize)
                     {
-                        throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(
-                            new ProtocolException(SR.HttpContentLengthIncorrect));
+                        throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(MaxMessageSizeStream.CreateMaxReceivedMessageSizeExceededException(_settings.MaxBufferSize));
                     }
 
-                    break;
-                }
-                count -= bytesRead;
-                offset += bytesRead;
-            }
+                    if (result.IsCompleted)
+                    {
+                        break;
+                    }
 
-            return await DecodeBufferedMessageAsync(new ArraySegment<byte>(buffer, 0, offset), inputStream);
-        }
+                    reader.AdvanceTo(buffer.GetPosition(0), buffer.End);
+                }
 
-        private async Task<Message> ReadChunkedBufferedMessageAsync(Stream inputStream)
-        {
-            try
-            {
-                return _messageEncoder.ReadMessage(await BufferMessageStreamAsync(inputStream, _bufferManager, _settings.MaxBufferSize), _bufferManager, ContentType);
+                message = await DecodeBufferedMessageAsync(buffer);
             }
-            catch (XmlException xmlException)
+            finally
             {
-                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(
-                    new ProtocolException(SR.MessageXmlProtocolError, xmlException));
+                reader.AdvanceTo(buffer.Start, buffer.End);
             }
```

Comment:
This line doesn't tell the reader it can release the buffer. The purpose of reader.AdvanceTo is to tell the reader what you have consumer so it can release the buffers that were allocated. This should be reader.AdvanceTo(buffer.End) to tell the reader you have consumed everything and it can return the buffers.

Reply:
@mconnew i think we might still have an issue as we do need the buffer in the DispatchAsync call which is made a bit later in AspNetCoreReplyChannel.HandleRequest. For the time being i'm delaying the buffer being freed by returning the PipeReader created by HttpInput.ParseIncomingMessageAsync to AspNetCoreReplyChannel.HandleRequest and call PipeReader.CompleteAsync there. 
I'm wondering if this is correct or not.. and i'm thinking to move the call to AdvanceTo(buffer.Start, buffer.End) after the DispatchAsync.